src/crimson/osd/osd_operation_sequencer.h

```diff
@@ -69,10 +69,40 @@ class OpSequencer {
   const ClientRequest* get_last_issued() const {
     return last_issued;
   }
-  void finish_op(const ClientRequest* const op) {
+  void finish_op_in_order(const ClientRequest* const op) {
     assert(op->get_id() > last_completed->get_id());
     last_completed = op;
   }
+  void finish_op_out_of_order(ClientRequest* const this_op) {
+    assert(this_op);
+    for (auto it = OperationRegistryI::op_list::s_iterator_to(*this_op);
+         it != OperationRegistryI::op_list::s_iterator_to(*last_issued);
+         ++it) {
+      // we're iterating over the operation registry of all ClientRequests.
+      // this list aggrates every single instance in the system, and thus
+      // we need to skip operations coming from different client's session
+      // or targeting different PG. The op we're interested in should be
+      // somewhere between the operation that early exists (this_op) and
+      // the most recently registered one.
+      // as the out-of-order exit is supposed to happen on cold paths only,
+      // the extra overhead is a thing we can live with.
```

Comment:
Could you elaborate a little bit about the "out-of-order exit"?

Reply:
IIUC, "out-of-order exit" can only happen when the op is read-only, am I right?

Reply:
Nope. Before the patches `OpSequencer` was assuming every single (inside a specific session / pg boundaries) finishes in the same exact order as it was entered. This isn't truly the case as we have a bunch of early exists like `send_incremenetal_map()` on misdirected op. See https://github.com/ceph/ceph/pull/41866.